tripy/tripy/constraints.py

```diff
@@ -78,29 +230,68 @@ def dtypes(
-            (e.g. __add__ and __radd__), this will enable type information to be added to the
+            (e.g. `__add__` and `__radd__`), this will enable type information to be added to the
             documentation for the aliases as well.
+
+        convert_tensor_and_shape_likes: If set to True, this decorator will also convert
+            arguments to `Tensor` or `DimensionSize`. If the argument can be converted to a
+            `DimensionSize`, it is. Otherwise, it is converted to a `Tensor`.
+
+            The conversions will respect any datatype constraints, casting the TensorLike values as necessary,
+            but will raise an exception for lossy casts like float to int (but *not* for, e.g., `float32` to `float16`).
+
+        conversion_targets: If `convert_tensor_and_shape_likes` is true, only the arguments named in the list
+            will be converted to `Tensor` or `DimensionSize`. If not supplied, any arguments annotated
+            with `TensorLike` or `ShapeLike` are converted if convert_tensor_and_shape_likes is true.
+
+        conversion_preprocess_func: If `convert_tensor_and_shape_likes` is true, this argument is a callback that is
+            used to preprocess the arguments before potential conversion. In this case, if provided, the callback
+            will be called regardless of whether the decorator performs any conversions.
+
+            The callback will be called with all arguments that were passed to the decorated function and should
+            return a dictionary of all updated arguments. For a variadic arg, the dictionary entry for the name
+            should have a list of all the updated values.
     """
 
     def decorator(func):
+        nonlocal conversion_targets
+
+        from tripy.types import ShapeLike, TensorLike
+
         return_dtype = constraints.get(RETURN_VALUE, None)
         VerifInfo = namedtuple("VerifInfo", ["obj", "inputs", "exceptions", "return_dtype", "dtypes", "constraints"])
         verif_info = VerifInfo(func, {}, exceptions, return_dtype, variables, constraints)
 
-        for key in [func.__qualname__] + aliases:
-            TYPE_VERIFICATION[key] = verif_info
+        signature = inspect.signature(func)
+        conversion_targets = conversion_targets or {
+            name for name, param in signature.parameters.items() if param.annotation in {TensorLike, ShapeLike}
+        }
+        shape_likes = {name for name, param in signature.parameters.items() if param.annotation is ShapeLike}
+
+        # if no dtype constraints have been specified at all, do not add to the table so we don't generate invalid tests
+        if constraints or variables or exceptions:
+            for key in [func.__qualname__] + aliases:
+                TYPE_VERIFICATION[key] = verif_info
 
         @functools.wraps(func)
         def wrapper(*args, **kwargs):
+            if convert_tensor_and_shape_likes:
+                args, kwargs = convert_input_types(
```

Comment:
Would be good if there's a way we can avoid doing `utils.merge_function_arguments` twice. Maybe we can call it here (conditionally, so we don't incur the cost if tensor conversion/dtype checking is disabled) and pass it as an input to `convert_input_types`? Then if `merged_args` is already set, we don't need to call it again in the next block.

Reply:
The main impediment is that we want to update the args during the type conversion and this would require also updating the merged args data structure. I will see if there will be an easy way to keep the merged args up to date.

Reply:
Sure, if it's non-trivial it's fine to keep it as-is.